src/c/device-coap.c

```diff
@@ -0,0 +1,225 @@
+/* Entry point for an EdgeX device service that provides a CoAP server to receive
+ * values asynchronously.
+ *
+ * Copyright (c) 2020
+ * Ken Bannister
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <unistd.h>
+#include <stdarg.h>
+
+#include "devsdk/devsdk.h"
+#include "device-coap.h"
+
+#define ERR_CHECK(x) if (x.code) { fprintf (stderr, "Error: %d: %s\n", x.code, x.reason); devsdk_service_free (service); free (impl); return x.code; }
+
+#define ERR_BUFSZ 1024
+#define COAP_BIND_ADDR_KEY "CoapBindAddr"
+#define SECURITY_MODE_KEY  "SecurityMode"
+#define PSK_KEY_KEY        "PskKey"
+
+
+/* Looks up security mode enum value from configuration text value */
+static coap_security_mode_t find_security_mode
+(
+  const char *mode_text
+)
+{
+  assert (mode_text);
+  if (!strcmp (mode_text, "PSK"))
+  {
+    return SECURITY_MODE_PSK;
+  }
+  else if (!strcmp (mode_text, "NoSec"))
+  {
+    return SECURITY_MODE_NOSEC;
+  }
+  else
+  {
+    return SECURITY_MODE_UNKNOWN;
+  }
+}
+
+/* Init callback; reads in config values to device driver */
+static bool coap_init
+(
+  void *impl,
+  struct iot_logger_t *lc,
+  const iot_data_t *config
+)
+{
+  coap_driver *driver = (coap_driver *) impl;
+
+  driver->lc = lc;
+  driver->security_mode = find_security_mode (iot_data_string_map_get_string (config, SECURITY_MODE_KEY));
+
+  if (driver->security_mode == SECURITY_MODE_PSK)
+  {
+    const char *conf_psk_key = iot_data_string_map_get_string (config, PSK_KEY_KEY);
+    if (strlen (conf_psk_key))
+    {
+      iot_data_t *key_array = iot_data_alloc_array_from_base64 (conf_psk_key);
+      driver->psk_key = key_array;
+      iot_log_info (lc, "PSK key len %u", iot_data_array_length (key_array));
+    }
+    else
+    {
+      iot_log_error (lc, "PSK key not in configuration");
+      driver->psk_key = NULL;
+      return false;
+    }
+  }
+  else
+  {
+    driver->psk_key = NULL;
+  }
+
+  /* CoAP server bind address as text */
+  const char *bind_addr = iot_data_string_map_get_string (config, COAP_BIND_ADDR_KEY);
+  if (bind_addr)
+  {
+    driver->coap_bind_addr = iot_data_alloc_string (bind_addr, IOT_DATA_COPY);
+  }
+  else
+  {
+    iot_log_error (lc, "CoAP bind address not in configuration");
+    driver->coap_bind_addr = NULL;
+    return false;
+  }
+
+  iot_log_debug (lc, "Init complete");
+  return true;
+}
+
+static bool coap_get_handler
+(
+  void *impl,
+  const char *devname,
+  const devsdk_protocols *protocols,
+  uint32_t nreadings,
+  const devsdk_commandrequest *requests,
+  devsdk_commandresult *readings,
+  const devsdk_nvpairs *qparms,
+  iot_data_t **exception
+)
+{
+  (void) impl;
+  (void) devname;
+  (void) protocols;
+  (void) nreadings;
+  (void) requests;
+  (void) readings;
+  (void) qparms;
+  (void) exception;
+  return true;
```

Comment:
Should return false and set `exception` to indicate the error, eg to `iot_data_alloc_string ("Request not supported; CoAP devices are push-only", IOT_DATA_REF)`

Reply:
Addressed in e55f995